aca.hpp

```diff
@@ -0,0 +1,90 @@
+#ifndef ACA_HPP
+#define ACA_HPP
+#include <unordered_map>
+#include <utility>
+#include <vector>
+#include <queue>
+
+template <typename CharT, typename DataT>
+struct TrieNode {
+  std::unordered_map<CharT, TrieNode<CharT, DataT> *> nexts;
+  TrieNode<CharT, DataT> * fail;
+  std::basic_string<CharT> str;
+  DataT data;
+  TrieNode () {
+    fail = nullptr;
+  }
+};
+
+template <typename CharT, typename DataT>
+struct Automation {
+  typedef TrieNode<CharT, DataT> NodeType;
+  NodeType * root_node = nullptr;
+  Automation () {
+    root_node = new NodeType();
+  }
+  ~Automation () {
+    destory(root_node);
+  }
+  void destory(NodeType *node){
+    for (auto it=node->nexts.begin(); it != node->nexts.end(); it++ ) {
+      destory(it->second);
+    }
+    delete node;
+  }
+  void insert(const std::basic_string<CharT> &str, const std::function<void(DataT &data)> &callback) {
+    NodeType * current = root_node;
+    for (auto chr = str.begin(); chr != str.end(); chr ++) {
+      auto next = current->nexts.find(*chr);
+      if (next == current->nexts.end()) {
+        NodeType *new_node = new NodeType();
+        current->nexts.insert(std::make_pair(*chr, new_node));
+        current = new_node;
+      } else {
+        current = next->second;
+      }
+    }
+    current->str = str;
```

Comment:
if only matching string has this field, it should remove it, because trienode has data field.